designs/drift.md

```diff
@@ -0,0 +1,81 @@
+
+(2) Dynamic Fields can correspond to multiple values and must be handled differently. Dynamic fields can create cases where drift occurs without changes to CRDs, or where CRD changes do not result in drift.
+
+For example, if a machine has `node.kubernetes.io/instance-type: m5.large`, and requirements change from `node.kubernetes.io/instance-type In [m5.large]` to `node.kubernetes.io/instance-type In [m5.large, m5.2xlarge]`, the machine will not be drifted because it's value is still compatible with the new requirements. Conversely, for an AWS Installation, if a machine is using a machine image `ami: ami-abc`, but a new image is published, Karpenter's `AWSNodeTemplate.amiSelector` will discover that the new correct value is `ami: ami-xyz`, and detect the machine as drifted.
+
+(3) Behavioral Fields are treated as over-arching settings on the Provisioner to dictate how Karpenter behaves. These fields don’t correspond to settings on the machine or instance. They’re set by the user to control Karpenter’s Provisioning and Deprovisioning logic. Since these don’t map to a desired state of machines, these fields will not be considered for Drift.
+
+Each of the currently defined fields in Karpenter's existing CRDs will be classified as follows (AWS Fields included here as well):
+
+```
+|                            | Static | Dynamic | Behavioral |
+|----------------------------|--------|---------|------------|
+| - Provisioner Fields -     |   ---  |   ---   |     ---    |
+| Startup Taints             |    x   |         |            |
+| Taints                     |    x   |         |            |
+| Labels                     |    x   |         |            |
+| Annotations                |    x   |         |            |
+| Node Requirements          |        |    x    |            |
+| Kubelet Configuration      |    x   |         |            |
+| Weight                     |        |         |      x     |
+| Limits                     |        |         |      x     |
+| Consolidation              |        |         |      x     |
+| TTLSecondsUntilExpired     |        |         |      x     |
+| TTLSecondsAfterEmpty       |        |         |      x     |
+| - AWSNodeTemplate Fields - |   ---  |   ---   |     ---    |
+| Subnet Selector            |        |    x    |            |
+| Security Group Selector    |        |    x    |            |
+| Instance Profile           |    x   |         |            |
+| AMI Family/AMI Selector    |        |    x    |            |
+| UserData                   |    x   |         |            |
+| Tags                       |    x   |         |            |
+| Metadata Options           |    x   |         |            |
+| Block Device Mappings      |    x   |         |            |
+| Detailed Monitoring        |    x   |         |            |
```

Comment:
Linking https://github.com/aws/karpenter/pull/3819 for @myaser to reflect on how the proposed design interacts with this drift implementation. I'm thinking that it should take a fairly similar shape, since we should be able to fully rely on the static data (since there is no discovery mechanism occurring with the network interfaces specified directly).